File: simpledb/SimpleDB.py
#!/usr/bin/env python
"""
Sample Python code for the O'Reilly book "Using AWS Infrastructure Services"
by James Murty.

This code was written for Python version 2.4.4 or greater, and requires the
XML/XPath library "lxml" version 1.3.6 or greater (see
http://codespeak.net/lxml/).

The SimpleDB module implements the Query API of the Amazon SimpleDB service.
"""

from AWS import AWS
import types
import time
import re

class SimpleDB(AWS):

  ENDPOINT_URI = 'https://sdb.amazonaws.com/'
  API_VERSION = '2007-11-07'
  SIGNATURE_VERSION = '1'

  HTTP_METHOD = 'POST' # 'GET'

  XML_NAMESPACE = 'http://sdb.amazonaws.com/doc/2007-11-07/'

  prior_box_usage = 0
  total_box_usage = 0


  def do_sdb_query(self, params):
    response = self.do_query(self.HTTP_METHOD, self.ENDPOINT_URI, params)
    xmldoc = self.parse_xml(response.read())

    self.prior_box_usage = float(self.xpath_value('//ns:BoxUsage', xmldoc))
    self.total_box_usage += self.prior_box_usage

    return xmldoc


  def list_domains(self, max_domains = 100):
    more_domains = True
    next_token = None
    domain_names = []

    while more_domains:
      params = self.build_query_parameters(
        self.API_VERSION, self.SIGNATURE_VERSION,
        {'Action' : 'ListDomains',
         'MaxNumberOfDomains' : max_domains,
         'NextToken' : next_token})

      xmldoc = self.do_sdb_query(params)

      for node in self.xpath_list('//ns:DomainName', xmldoc):
        domain_names.append(node.text)

      # If we receive a NextToken element, perform a follow-up operation
      # to retrieve the next set of domain names.
      next_token = self.xpath_value('//ns:NextToken', xmldoc)
      more_domains = next_token != None

    return domain_names


  def create_domain(self, domain_name):
    params = self.build_query_parameters(
      self.API_VERSION, self.SIGNATURE_VERSION,
      {'Action' : 'CreateDomain',
       'DomainName' : domain_name})

    self.do_sdb_query(params)
    return True


  def delete_domain(self, domain_name):
    params = self.build_query_parameters(
      self.API_VERSION, self.SIGNATURE_VERSION,
      {'Action' : 'DeleteDomain',
       'DomainName' : domain_name
      })

    self.do_sdb_query(params)
    return True


  def build_attribute_params(self, attributes={}, replace=False):
    attribute_params = {}
    index = 0

    for attrib_name in attributes:
      attrib_value = attributes[attrib_name]

      if type(attrib_value) != types.ListType:
        attrib_value = [attrib_value]

      for value in attrib_value:
        attribute_params['Attribute.%i.Name' % index] = attrib_name
        if value != None:
          # Automatically encode attribute values
          value = self.encode_attribute_value(value)
          attribute_params['Attribute.%i.Value' % index] = value

        # Add a Replace parameter for the attribute if the replace flag is set
        if replace:
          attribute_params['Attribute.%i.Replace' % index] = 'true'
        index += 1

    return attribute_params


  def get_attributes(self, domain_name, item_name, attribute_name = None):
    params = self.build_query_parameters(
      self.API_VERSION, self.SIGNATURE_VERSION,
      {'Action' : 'GetAttributes',
       'DomainName' : domain_name,
       'ItemName' : item_name,
       'AttributeName' : attribute_name})

    xmldoc = self.do_sdb_query(params)

    attributes = {}
    for attribute_node in self.xpath_list('//ns:Attribute', xmldoc):
      attr_name = self.xpath_value('ns:Name', attribute_node)
      value = self.xpath_value('ns:Value', attribute_node)

      # Automatically decode attribute values
      value = self.decode_attribute_value(value)

      # An empty attribute value is an empty string, not nil.
      if not value:
        value = ''

      if attr_name in attributes:
        attributes[attr_name].append(value)
      else:
        attributes[attr_name] = [value]

    if attribute_name:
      # If a specific attribute was requested, return only the values array
      # for this attribute.
      if not attribute_name in attributes:
        return []
      else:
        return attributes[attribute_name]
    else:
      return attributes


  def put_attributes(self, domain_name, item_name, attributes, replace=False):
    params = self.build_query_parameters(
      self.API_VERSION, self.SIGNATURE_VERSION,
      {'Action' : 'PutAttributes',
       'DomainName' : domain_name,
       'ItemName' : item_name})

    attrib_params = self.build_attribute_params(attributes, replace)
    for param_name in attrib_params:
      params[param_name] = attrib_params[param_name]

    self.do_sdb_query(params)
    return True


  def delete_attributes(self, domain_name, item_name, attributes={}):
    params = self.build_query_parameters(
      self.API_VERSION, self.SIGNATURE_VERSION,
      {'Action' : 'DeleteAttributes',
       'DomainName' : domain_name,
       'ItemName' : item_name})

    attrib_params = self.build_attribute_params(attributes)
    for param_name in attrib_params:
      params[param_name] = attrib_params[param_name]

    self.do_sdb_query(params)
    return True


  def query(self, domain_name, query_expression = None,
            max_items = 100, fetch_all = True):
    more_items = True
    next_token = None
    item_names = []

    while more_items:
      params = self.build_query_parameters(
        self.API_VERSION, self.SIGNATURE_VERSION,
        {'Action' : 'Query',
         'DomainName' : domain_name,
         'QueryExpression' : query_expression,
         'MaxNumberOfItems' : max_items,
         'NextToken' : next_token})

      xmldoc = self.do_sdb_query(params)

      for item_name_node in self.xpath_list('//ns:ItemName', xmldoc):
        item_names.append(item_name_node.text)

      next_token = self.xpath_value('//ns:NextToken', xmldoc)
      more_items = fetch_all and next_token

    return item_names


  def encode_boolean(self, value):
    if value:
      return '!b'
    else:
      return '!B'


  def decode_boolean(self, value_str):
    if value_str == '!B':
      return False
    elif value_str == '!b':
      return True
    else:
      raise 'Cannot decode boolean from string: %s' % value_str


  def encode_date(self, value):
    return "!d" + time.strftime(self.ISO8601, value)


  def decode_date(self, value_str):
    if value_str[0:2] == '!d':
      return time.strptime(value_str[2:], self.ISO8601)
    else:
      raise 'Cannot decode date from string: %s' % value_str


  def encode_integer(self, value, max_digits=18):
    upper_bound = (10 ** max_digits)

    if value >= upper_bound or value < -upper_bound:
      raise 'Integer %i is outside encoding range (-%i to %i)' % \
        (value, upper_bound, (upper_bound - 1))

    format = '%%0%dd' % max_digits
    if value < 0:
      return "!I" + format % (upper_bound + value)
    else:
      return "!i" + format % value


  def decode_integer(self, value_str):
    if value_str[0:2] == '!I':
      # Encoded value is a negative integer
      max_digits = len(value_str) - 2
      upper_bound = (10 ** max_digits)

      return int(value_str[2:]) - upper_bound
    elif value_str[0:2] == '!i':
      # Encoded value is a positive integer
      return int(value_str[2:])
    else:
      raise 'Cannot decode integer from string: %s' % value_str


  def encode_float(self, value, max_exp_digits=2, max_precision_digits=15):
    exp_midpoint = (10 ** max_exp_digits) / 2

    # This is a horrible string-based hack to work out the fraction and
    # exponent comoponents of a float.
    str = ('%%.%ie' % (max_precision_digits - 1)) % value
    if value >= 0:
      fraction = str[0] + str[2:(max_precision_digits + 1)]
      exponent = int(str[(max_precision_digits + 2):])
      if exponent != 0:
        exponent += 1
    else:
      fraction = str[1] + str[3:(max_precision_digits + 2)]
      exponent = int(str[(max_precision_digits + 3):]) + 1

    if exponent >= exp_midpoint or exponent < -exp_midpoint:
      raise 'Exponent %i is outside encoding range (-%i to %i)' % \
        (exponent, exp_midpoint, (exp_midpoint - 1))

    # The zero value is a special case, for which the exponent must be 0
    if value == 0:
      exponent = -exp_midpoint

    exp_format = '%%0%dd' % max_exp_digits
    fraction_format = '%%0%dd' % max_precision_digits

    if value >= 0:
      return '!f%s!%s' % ((exp_format % (exp_midpoint + exponent)), fraction)
    else:
      fraction_upper_bound = (10 ** max_precision_digits)
      diff_fraction = fraction_upper_bound - int(fraction)
      return '!F%s!%s' % (exp_format % (exp_midpoint - exponent),
        fraction_format % diff_fraction)


  def decode_float(self, value_str):
    prefix = value_str[0:2]

    if prefix != '!f' and prefix != '!F':
      raise 'Cannot decode float from string: %s' % value_str

    matches = re.findall(r'![fF]([0-9]+)!([0-9]+)', value_str)
    exp_str, fraction_str = matches[0]

    max_exp_digits = len(exp_str)
    exp_midpoint = (10 ** max_exp_digits) / 2
    max_precision_digits = len(fraction_str)

    if prefix == '!F':
      sign = -1
      exp = exp_midpoint - int(exp_str)

      fraction_upper_bound = (10 ** max_precision_digits)
      fraction = fraction_upper_bound - int(fraction_str)
    else:
      sign = 1
      exp = int(exp_str) - exp_midpoint

      fraction = int(fraction_str)

    return sign * float('0.%i' % fraction) * (10 ** exp)


  def encode_attribute_value(self, value):
    if value == True or value == False:
      return self.encode_boolean(value)
    elif type(value) == time.struct_time:
      return self.encode_date(value)
    elif type(value) == types.IntType:
      return self.encode_integer(value)
    elif type(value) == types.FloatType:
      return self.encode_float(value)
    else:
      # No type-specific encoding is available, so we simply convert
      # the value to a string.
      return str(value)


  def decode_attribute_value(self, value_str):
    if not value_str:
      return ''

    # Check whether the '!' flag is present to indicate an encoded value
    if value_str[0] != '!':
      return value_str

    prefix = value_str[0:2].lower()
    if prefix == '!b':
      return self.decode_boolean(value_str)
    elif prefix == '!d':
      return self.decode_date(value_str)
    elif prefix == '!i':
      return self.decode_integer(value_str)
    elif prefix == '!f':
      return self.decode_float(value_str)
    else:
      return value_str


if __name__ == '__main__':
  sdb = SimpleDB(debug_mode=True)

  print sdb.list_domains()

  #print sdb.create_domain('test-domain')

  #attribs = \
  #  {'Name' : 'Tomorrow Never Knows',
  #   'Artist' : 'The Beatles',
  #   'Time' : '177',
  #   'Album' : ['Revolver','The Beatles Box Set']}
  #sdb.put_attributes('test-domain', 'TestItem', attribs)

  #print sdb.put_attributes('test-domain', 'TestItem', {'Time' : '2:57'})

  #print sdb.put_attributes('test-domain', 'TestItem', {'Time' : '2:57'}, True)

  #print sdb.get_attributes('test-domain', 'TestItem')

  #print sdb.get_attributes('test-domain', 'TestItem', 'Time')

  #print sdb.delete_attributes('test-domain', 'TestItem', {'Album' : 'Revolver'})

  #print sdb.delete_attributes('test-domain', 'TestItem', {'Album' : None})

  #print sdb.delete_attributes('test-domain', 'TestItem')

  #results = sdb.query('stocks', "['Code' = 'AAPL']", max_items = 250)
  #print results
  #print len(results)

  #results = sdb.query('stocks', "['Date' > '2007-06-05T00:00:00Z']")
  #print results
  #print len(results)

  #print sdb.delete_domain('test-domain')

  #print 'Prior usage: %f' % sdb.prior_box_usage
  #print 'Total usage: %f' % sdb.total_box_usage

  #print 'Encoded booleans: True=%s, False=%s' % \
  #  (sdb.encode_boolean(True), sdb.encode_boolean(False))
  #print 'Decoded booleans: !b=%s, !B=%s' % \
  #  (sdb.decode_boolean('!b'), sdb.decode_boolean('!B'))

  #enc_date = sdb.encode_date(time.gmtime(time.time()))
  #print 'Encoded Date: %s' % enc_date
  #print 'Decoded Date: %s' % sdb.decode_date(enc_date)

  #max_digits = 2
  #print sdb.encode_integer(7, max_digits)
  #print sdb.decode_integer('!i07')
  #print sdb.encode_integer(25, max_digits)
  #print sdb.decode_integer('!i25')
  #print sdb.encode_integer(-3, max_digits)
  #print sdb.decode_integer('!I97')
  #print sdb.encode_integer(-100, max_digits)
  #print sdb.decode_integer('!I00')

  #print sdb.encode_float(0.0)
  #print sdb.decode_float('!f50!000000000000000')
  #print sdb.encode_float(12345678901234567890)
  #print '%15f' % sdb.decode_float('!f70!123456789012346')
  #print sdb.encode_float(0.12345678901234567890)
  #print '%.15f' % sdb.decode_float('!f50!123456789012346')
  #print sdb.encode_float(-12345678901234567890)
  #print '%15f' % sdb.decode_float('!F30!876543210987654')
  #print sdb.encode_float(-0.12345678901234567890)
  #print '%.15f' % sdb.decode_float('!F50!876543210987654')

  #print sdb.encode_attribute_value(False)
  #print sdb.decode_attribute_value('!B')
  #print sdb.encode_attribute_value(time.gmtime())
  #print sdb.decode_attribute_value('!d2008-01-25T05:32:39Z')
  #print sdb.encode_attribute_value(43.54)
  #print sdb.decode_attribute_value('!f52!435400000000000')
  #print sdb.encode_attribute_value(8637)
  #print sdb.decode_attribute_value('!i000000000000008637')

File: java/com/oreilly/aws/SimpleDB.java
package com.oreilly.aws;

import java.io.ByteArrayOutputStream;
import java.io.PrintStream;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

import org.w3c.dom.Document;
import org.w3c.dom.Node;

/**
 * Sample Java code for the O'Reilly book "Using AWS Infrastructure Services"
 * by James Murty.
 * <p>
 * This code was written for Java version 5.0 or greater. If you are not using
 * Sun's Java implementation, this also code requires the Apache Commons Codec
 * library (see http://commons.apache.org/codec/)
 * <p>
 * The SimpleDB class implements the Query API of the Amazon SimpleDB service.
 */
public class SimpleDB extends AWS {

    public static URL ENDPOINT_URI;
    public static final String API_VERSION = "2007-11-07";
    public static final String SIGNATURE_VERSION = "1";

    public HttpMethod HTTP_METHOD = HttpMethod.POST; // GET

    protected double priorBoxUsage = 0.0;
    protected double totalBoxUsage = 0.0;


    static {
        try {
            ENDPOINT_URI = new URL("https://sdb.amazonaws.com/");
        } catch (MalformedURLException e) {
            e.printStackTrace();
        }
    }

    /**
     * Initialize the service and set the service-specific variables:
     * awsAccessKey, awsSecretKey, isDebugMode, and isSecureHttp.
     *
     * This constructor obtains your AWS access and secret key credentials from
     * the AWS_ACCESS_KEY and AWS_SECRET_KEY environment variables respectively.
     * It sets isDebugMode to false, and isSecureHttp to true.
     */
    public SimpleDB() {
        super();
    }

    /**
     * Initialize the service and set the service-specific variables:
     * awsAccessKey, awsSecretKey, isDebugMode, and isSecureHttp.
     *
     * This constructor obtains your AWS access and secret key credentials from
     * the AWS_ACCESS_KEY and AWS_SECRET_KEY environment variables respectively.
     * It sets isDebugMode and isSecureHttp according to the values you provide.
     */
    public SimpleDB(boolean isDebugMode, boolean isSecureHttp) {
        super(isDebugMode, isSecureHttp);
    }

    /**
     * Initialize the service and set the service-specific variables:
     * awsAccessKey, awsSecretKey, isDebugMode, and isSecureHttp.
     */
    public SimpleDB(String awsAccessKey, String awsSecretKey,
        boolean isDebugMode, boolean isSecureHttp)
    {
        super(awsAccessKey, awsSecretKey, isDebugMode, isSecureHttp);
    }

    public double getPriorBoxUsage() {
        return priorBoxUsage;
    }

    public double getTotalBoxUsage() {
        return totalBoxUsage;
    }

    /**
     * Uses the doQuery method defined in AWS to sends a GET or POST request
     * message to the SimpleDB service's Query API interface and returns the
     * response result from the service.
     *
     * This method performs retrieves box usage values from the service's
     * response.
     */
    protected Document doSdbQuery(Map<String, String> parameters)
        throws Exception
    {
        parameters = buildQueryParameters(API_VERSION, SIGNATURE_VERSION, 
            parameters, EMPTY_INDEXED_MAP);        
        HttpURLConnection conn = doQuery(HTTP_METHOD, ENDPOINT_URI, parameters);
        Document xmlDoc = parseToDocument(conn.getInputStream());
        
        String usageStr = xpathToContent("//BoxUsage", xmlDoc);
        if (usageStr != null) {
            priorBoxUsage = Double.parseDouble(usageStr);
            totalBoxUsage += priorBoxUsage;
        }

        return xmlDoc;
    }
    
    public List<String> listDomains() throws Exception {
        return listDomains(100);
    }
    
    public List<String> listDomains(int maxDomains) throws Exception {
        boolean moreDomains = true;
        String nextToken = null;
        List<String> domainNames = new ArrayList<String>();
                        
        while (moreDomains) {        
            Map<String, String> parameters = new HashMap<String, String>();
            parameters.put("Action", "ListDomains");
            parameters.put("MaxNumberOfDomains", String.valueOf(maxDomains));
            parameters.put("NextToken", nextToken);
                
            Document xmlDoc = doSdbQuery(parameters);

            for (Node node : xpathToNodeList("//DomainName", xmlDoc)) {
                domainNames.add(node.getTextContent());
            }
            
            // If we receive a NextToken element, perform a follow-up operation
            // to retrieve the next set of domain names.
            nextToken = xpathToContent("//NextToken/text()", xmlDoc);
            moreDomains = (nextToken != null);
        }
        return domainNames;
    }
    
    public boolean createDomain(String domainName) throws Exception {
        Map<String, String> parameters = new HashMap<String, String>();
        parameters.put("Action", "CreateDomain");
        parameters.put("DomainName", domainName);
    
        doSdbQuery(parameters);
        return true;
    }
    
    public boolean deleteDomain(String domainName) throws Exception {
        Map<String, String> parameters = new HashMap<String, String>();
        parameters.put("Action", "DeleteDomain");
        parameters.put("DomainName", domainName);
    
        doSdbQuery(parameters);
        return true;
    }
    
    protected Map<String, String> buildAttributeParams(
        Map<String, Object> attributes, boolean replace) throws Exception 
    {
        Map<String, String> attributeParams = new HashMap<String, String>();
        int index = 0;
        
        if (attributes == null) {
            return attributeParams;
        }
        
        for (Map.Entry<String, Object> param : attributes.entrySet()) {
            List valueList = null;
            if (param.getValue() instanceof List) {
                valueList = (List) param.getValue();
            }  else {
                valueList = new ArrayList();
                valueList.add(param.getValue());
            }
            
            for (Object value: valueList) {
                attributeParams.put("Attribute." + index + ".Name", param.getKey());
                if (value != null) {
                    // Automatically encode attribute values
                    String encodedValue = encodeAttributeValue(value);
                    attributeParams.put("Attribute." + index + ".Value", 
                        encodedValue);                    
                }
                // Add a Replace parameter for the attribute if the replace flag is set
                if (replace) {
                    attributeParams.put("Attribute." + index + ".Replace", "true");
                }
                index++;
            }
        }        
        return attributeParams;
    }
    
    public boolean putAttributes(String domainName, String itemName, 
        Map<String, Object> attributes) throws Exception
    {
        return putAttributes(domainName, itemName, attributes, false);
    }

    public boolean putAttributes(String domainName, String itemName, 
        Map<String, Object> attributes, boolean replace) throws Exception 
    {
        Map<String, String> parameters = new HashMap<String, String>();
        parameters.put("Action", "PutAttributes");
        parameters.put("DomainName", domainName);
        parameters.put("ItemName", itemName);
        
        parameters.putAll(buildAttributeParams(attributes, replace));
    
        doSdbQuery(parameters);
        return true;
    }

    public boolean deleteAttributes(String domainName, String itemName) 
        throws Exception 
    {
        return deleteAttributes(domainName, itemName, null);
    }
    
    public boolean deleteAttributes(String domainName, String itemName, 
        Map<String, Object> attributes) throws Exception 
    {
        Map<String, String> parameters = new HashMap<String, String>();
        parameters.put("Action", "DeleteAttributes");
        parameters.put("DomainName", domainName);
        parameters.put("ItemName", itemName);
        
        parameters.putAll(buildAttributeParams(attributes, false));
    
        doSdbQuery(parameters);
        return true;
    }
    
    public Map<String, List> getAttributes(String domainName, String itemName) 
        throws Exception
    {
        return getAttributesImpl(domainName, itemName, null);
    }

    
    public List getAttributes(String domainName, String itemName, 
        String attributeName) throws Exception
    {
        // When a specific attribute is requested, return only the values array
        // list for this attribute.
        Map<String, List> attributes = 
            getAttributesImpl(domainName, itemName, attributeName);
        List valueList = (List) attributes.get(attributeName);
        if (valueList == null) {
            return new ArrayList();
        } else {
            return valueList;
        }
    }

    protected Map<String, List> getAttributesImpl(String domainName, String itemName, 
        String attributeName) throws Exception 
    {
        Map<String, String> parameters = new HashMap<String, String>();
        parameters.put("Action", "GetAttributes");
        parameters.put("DomainName", domainName);
        parameters.put("ItemName", itemName);
        parameters.put("AttributeName", attributeName);
        
        Document xmlDoc = doSdbQuery(parameters);
        
        Map<String, List> attributes = new TreeMap<String, List>();
        for (Node node : xpathToNodeList("//Attribute", xmlDoc)) {
            String name = xpathToContent("Name", node);
            String value = xpathToContent("Value", node);
            
            // Automatically decode attribute values
            Object decodedValue = decodeAttributeValue(value);
            
            if (decodedValue == null) {
                // An empty attribute value is an empty string, not null.
                decodedValue = "";
            }
            
            List valueList = attributes.get(name);
            if (valueList != null) {
                valueList.add(decodedValue);
            } else {
                valueList = new ArrayList();
                valueList.add(decodedValue);
                attributes.put(name, valueList);
            }
        }        
        return attributes;
    }

    public List<String> query(String domainName, String queryExpression) 
        throws Exception 
    {
        return query(domainName, queryExpression, 100, true);
    }
    
    public List<String> query(String domainName, String queryExpression,
        int maxItems, boolean fetchAll) throws Exception 
    {
        boolean moreItems = true;
        String nextToken = null;
        
        List<String> itemNames = new ArrayList<String>();
        while (moreItems) {        
            Map<String, String> parameters = new HashMap<String, String>();
            parameters.put("Action", "Query");
            parameters.put("DomainName", domainName);
            parameters.put("QueryExpression", queryExpression);
            parameters.put("NextToken", nextToken);
            parameters.put("MaxNumberOfItems", String.valueOf(maxItems));
        
            Document xmlDoc = doSdbQuery(parameters);
        
            for (Node node : xpathToNodeList("//ItemName", xmlDoc)) {
                itemNames.add(node.getTextContent());
            }
            
            nextToken = xpathToContent("//NextToken", xmlDoc);
            moreItems = (nextToken != null && fetchAll);
        }        
        return itemNames;
    }

    
    public String encodeBoolean(boolean value) {
        if (value) {
            return "!b";
        } else {
            return "!B";            
        }
    }
    
    public boolean decodeBoolean(String value) throws Exception {
        if ("!B".equals(value)) {
            return false;
        } else if ("!b".equals(value)) {
            return true;
        } else {
            throw new Exception("Cannot decode boolean from string: " + value);
        }
    }
    
    public String encodeDate(Date value) {
        return "!d" + iso8601DateFormat.format(value);
    }
    
    public Date decodeDate(String value) throws Exception {
        if (value.startsWith("!d")) {
            return iso8601DateFormat.parse(value.substring(2));            
        } else {
            throw new Exception("Cannot decode date from string: " + value);
        }        
    }
    
    protected String printf(String format, Object value) throws Exception {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        new PrintStream(baos).printf("%0" + 2 + "d", value);
        return new String(baos.toByteArray(), "UTF-8");
    }

    public String encodeInteger(long value) throws Exception {
        return encodeInteger(value, 18);
    }
    
    public String encodeInteger(long value, int maxDigits) throws Exception {
        long upperBound = Math.round(Math.pow(10, maxDigits));
        
        if (value > upperBound || value < -upperBound) {
            throw new Exception("Integer " + value + " is outside encoding range (-"
                + upperBound + " to " + (upperBound - 1));
        }
        
        if (value < 0) {
            return "!I" + printf("%0" + maxDigits + "d", upperBound + value);    
        } else {
            return "!i" + printf("%0" + maxDigits + "d", value);    
        }
    }
    
    public long decodeInteger(String value) throws Exception {
        if (value.startsWith("!I")) {
            // Encoded value is a negative integer
            int maxDigits = value.length() - 2;
            long upperBound = Math.round(Math.pow(10, maxDigits));

            return Long.valueOf(value.substring(2)) - upperBound;
        }  else if (value.startsWith("!i")) {
            // Encoded value is a positive integer
            return Long.valueOf(value.substring(2));
        } else {
            throw new Exception("Cannot decode integer from string: " + value);
        }        
    }

    public String encodeFloat(double value) throws Exception 
    {
        return encodeFloat(value, 2, 15);
    }

    public String encodeFloat(double value, int maxExpDigits, 
        int maxPrecisionDigits) throws Exception 
    {
        long expMidpoint = Math.round(Math.pow(10, maxExpDigits) / 2);
        
        String precisionFormat = "";
        for (int i = 0; i < maxPrecisionDigits; i++) {
            precisionFormat += "0";
        }              
        DecimalFormat formatter = new DecimalFormat(
            " ." + precisionFormat + "E0;-." + precisionFormat + "E0");
               
        String formattedValue = formatter.format(value);
        String fractionStr = formattedValue.substring(2, 2 + maxPrecisionDigits);
        int exponent = Integer.valueOf(formattedValue.substring(3 + maxPrecisionDigits));
        
        if (exponent >= expMidpoint || exponent < -expMidpoint) {
          throw new Exception ("Exponent " + exponent + 
              " is outside encoding range (-" + expMidpoint + " to " +
              (expMidpoint - 1));
        }
        
        if (value == 0) {
            // The zero value is a special case, for which the exponent must be 0
            exponent = (int) -expMidpoint;
        }

        if (value >= 0) {
            return "!f" + printf("%0" + maxExpDigits + "d", expMidpoint + exponent) +
                "!" + fractionStr;
        } else {
            long fractionUpperBound = Math.round(Math.pow(10, maxPrecisionDigits));
            long diffFraction = fractionUpperBound - Long.valueOf(fractionStr);
            return "!F" + printf("%0" + maxExpDigits + "d", expMidpoint - exponent) +
                "!" + printf("%0" + maxPrecisionDigits + "d", diffFraction);
        }
    }
    
    public double decodeFloat(String value) throws Exception {
        if (!value.startsWith("!f") && !value.startsWith("!F")) {
            throw new Exception("Cannot decode float from string: " + value);
        }
        
        int expIndex = value.substring(2).indexOf("!");
        String expStr = value.substring(2, 2 + expIndex);
        String fractionStr = value.substring(expIndex + 3, value.length());
        
        int maxExpDigits = expStr.length();
        long expMidpoint = Math.round(Math.pow(10, maxExpDigits) / 2);
        int maxPrecisionDigits = fractionStr.length();
        
        long fraction = 0;
        long exp = 0;
        int sign = 1;
        if (value.startsWith("!F")) {
            sign = -1;
            exp = expMidpoint - Long.valueOf(expStr);
            long fractionUpperBound = Math.round(Math.pow(10, maxPrecisionDigits));
            fraction = fractionUpperBound - Long.valueOf(fractionStr);
        } else {
            sign = 1;
            exp = Long.valueOf(expStr) - expMidpoint;
            fraction = Long.valueOf(fractionStr);
        }

       return sign * Double.valueOf("0." + fraction) * Math.pow(10, exp);
    }

    
    public String encodeAttributeValue(Object value) throws Exception {
      if (value == null) {
          return null;
      } else if (value instanceof Boolean) {
          return encodeBoolean((Boolean) value);
      } else if (value instanceof Date) {
          return encodeDate((Date) value);
      } else if (value instanceof Long) {
          return encodeInteger((Long) value);
      } else if (value instanceof Integer) {
          return encodeInteger((Integer) value);          
      } else if (value instanceof Short) {
          return encodeInteger((Short) value);          
      } else if (value instanceof Double) {
          return encodeFloat((Double) value);
      } else if (value instanceof Float) {
          return encodeFloat((Float) value);
      } else {
          // No type-specific encoding is available, so we simply convert
          // the value to a string.
          return String.valueOf(value);
      }
    }

    public Object decodeAttributeValue(String value) throws Exception {
        if (value == null) {
            return "";
        }
                
        // Check whether the '!' flag is present to indicate an encoded value
        if (!value.startsWith("!")) {
            return value;
        }
        
        String prefix = value.substring(0, 2).toUpperCase();
        if (prefix.startsWith("!B")) {
            return decodeBoolean(value);
        } else if (prefix.startsWith("!D")) {
            return decodeDate(value);
        } else if (prefix.startsWith("!I")) {
            return decodeInteger(value);
        } else if (prefix.startsWith("!F")) {
            return decodeFloat(value);
        } else {
            return value;
        }
    }
        
}
